cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_integral_dpc.cpp

```diff
@@ -0,0 +1,131 @@
+                const std::int64_t sg_id = sg.get_group_id()[0];
+                const std::int64_t wg_id = item.get_global_id(1);
+                const std::int64_t wg_count = item.get_global_range(1);
+                const std::int64_t sg_count = sg.get_group_range()[0];
+                const std::int64_t sg_global_id = wg_id * sg_count + sg_id;
+                const std::int64_t total_sg_count = wg_count * sg_count;
+
+                const std::int64_t local_id = sg.get_local_id()[0];
+                const std::int64_t local_range = sg.get_local_range()[0];
+
+                const std::int64_t block_size =
+                    row_count / total_sg_count + std::int64_t(row_count % total_sg_count > 0);
+                const std::int64_t offset = block_size * sg_global_id;
+                const std::int64_t end =
+                    (offset + block_size) > row_count ? row_count : (offset + block_size);
+                for (std::int64_t i = offset + local_id; i < end; i += local_range) {
+                    const std::int32_t cl = label_ptr[i];
+                    sycl::ONEAPI::atomic_ref<std::int32_t,
+                                             cl::sycl::ONEAPI::memory_order::relaxed,
+                                             cl::sycl::ONEAPI::memory_scope::device,
+                                             cl::sycl::access::address_space::global_device_space>
+                        counter_atomic(counter_ptr[cl]);
+                    counter_atomic.fetch_add(1);
+                }
+            });
+    });
+}
+
+sycl::event count_empty_clusters(sycl::queue& queue,
+                                 std::int64_t cluster_count,
+                                 pr::ndview<std::int32_t, 1>& counters,
+                                 pr::ndarray<std::int32_t, 1>& empty_cluster_count,
+                                 const bk::event_vector& deps) {
+    ONEDAL_ASSERT(counters.get_dimension(0) == cluster_count);
+    ONEDAL_ASSERT(empty_cluster_count.get_dimension(0) == 1);
+    const std::int32_t* counter_ptr = counters.get_data();
+    const auto sg_size_to_set = get_recommended_sg_size2(queue);
+    std::int32_t* value_ptr = empty_cluster_count.get_mutable_data();
+    return queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for<merge_counters>(
+            bk::make_multiple_nd_range_2d({ sg_size_to_set, 1 }, { sg_size_to_set, 1 }),
+            [=](sycl::nd_item<2> item) {
+                auto sg = item.get_sub_group();
+                const std::int64_t sg_id = sg.get_group_id()[0];
+                if (sg_id > 0)
+                    return;
+                const std::int64_t local_id = sg.get_local_id()[0];
+                const std::int64_t local_range = sg.get_local_range()[0];
+                std::int32_t sum = 0;
+                for (std::int64_t i = local_id; i < cluster_count; i += local_range) {
+                    sum += counter_ptr[i] == 0;
```

Comment:
sum and value_ptr are int32, but cluster_count is int64, in case if cluster count value is over int32max and each counter_ptr[i] eq 1, you will get int32_t overflow. Need to add a check here that cluster_count le int32max.

Reply:
Added check for cluster count